web-api/src/ingestion/routes.rs

```diff
@@ -229,56 +229,77 @@ async fn upsert_documents(
+    .map(|document| {
+        (
+            document.id,
+            (
+                document.snippet,
+                document.properties,
+                document.tags,
+                document.is_candidate,
+            ),
+        )
+    })
     .collect::<HashMap<_, _>>();
 
     // Hint: Documents which have a changed snippet are also in `new_documents`.
-    let (changed_documents, new_documents) = documents
+    let (new_documents, changed_documents) = documents
         .into_iter()
-        .map(|document| {
-            let (snippet, is_candidate) = existing_data
+        .partition_map::<Vec<_>, Vec<_>, _, _, _>(|document| {
+            let (data, is_candidate) = existing_documents
                 .get(&document.id)
-                .map(|(snippet, is_candidate)| (snippet, *is_candidate))
+                .map(|(snippet, properties, tags, is_candidate)| {
+                    ((snippet, properties, tags), *is_candidate)
+                })
                 .unzip();
 
+            let new_snippet = data.map_or(true, |(snippet, _, _)| snippet != &document.snippet);
             let new_is_candidate = document.is_candidate_op.resolve(is_candidate);
-            let unchanged_snippet = snippet
-                .map(|snippet| snippet == &document.snippet)
-                .unwrap_or_default();
 
-            (document, new_is_candidate, unchanged_snippet)
-        })
-        .partition::<Vec<_>, _>(|(_, _, unchanged_snippet)| *unchanged_snippet);
+            if new_snippet {
+                Either::Left((document, new_is_candidate))
+            } else {
+                let new_properties = data.map_or(true, |(_, properties, _)| {
+                    properties != &document.properties
+                });
+                let new_tags = data.map_or(true, |(_, _, tags)| tags != &document.tags);
+                Either::Right((document, new_properties, new_tags, new_is_candidate))
+            }
+        });
 
     storage::DocumentCandidate::remove(
         &state.storage,
         changed_documents
             .iter()
-            .filter_map(|(document, new_is_candidate, _)| {
+            .filter_map(|(document, _, _, new_is_candidate)| {
                 new_is_candidate
                     .has_changed_to_false()
                     .then_some(&document.id)
             }),
     )
     .await?;
 
-    for (document, _, _) in &changed_documents {
-        storage::DocumentProperties::put(&state.storage, &document.id, &document.properties)
-            .await?;
-        storage::Tag::put(&state.storage, &document.id, &document.tags).await?;
+    for (document, new_properties, new_tags, _) in &changed_documents {
+        if *new_properties {
+            storage::DocumentProperties::put(&state.storage, &document.id, &document.properties)
+                .await?;
+        }
+        if *new_tags {
+            storage::Tag::put(&state.storage, &document.id, &document.tags).await?;
+        }
```

Comment:
i would assume that it's faster in a single call instead of two. we would need to extend the storage api though, maybe in another pr if this becomes a problem. with the current setup, changed docs should be a rare case, i'd assume most docs are either new or unchanged for most customers using the ingestion endpoint. i hope if a customer wants to explicitely change properties or tags, that they use the appropriate endpoints instead of the ingestion endpoint.